In [ ]:
from math import factorial, exp
import numpy as np

In [ ]:
def combinations(n, k):
    return int(factorial(n) / (factorial(k) * factorial(n - k)))

# Домашнее задание к уроку 3


## Задание 1

Даны значения зарплат из выборки выпускников: 100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 24, 57, 55, 70, 75, 65, 84, 90, 150. Посчитать (желательно без использования статистических методов наподобие std, var, mean) среднее арифметическое, среднее квадратичное отклонение, смещенную и несмещенную оценки дисперсий для данной выборки.

In [ ]:
# выборка зарплат выпускников
salary = np.array([100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 24, 57, 55, 70, 75, 65, 84, 90, 150])

In [ ]:
# вычисление среднего арифметического
average = np.sum(salary) / len(salary)
average

65.3

In [ ]:
# среднее квадратичное отклонение
deviation = round(np.sqrt(np.sum((salary - average)**2) / len(salary)), 2)
deviation

30.82

In [ ]:
# смещённая оценка дисперсии
dispersion = np.sum((salary - average)**2) / len(salary)
dispersion


950.11

In [ ]:
# несмещённая оценка дисперсии
dispersion_non =round(np.sum((salary - average)**2) / (len(salary) - 1), 2)
dispersion_non

1000.12

**Ответ**: Среднее арифметическое значение данной выборки равно 65.3, среднее квадратичное отклонение - 30.82, смещённая оценка дисперсии - 950.11, несмещённая оценка дисперсии - 1000.12

## Задание 2

В первом ящике находится 8 мячей, из которых 5 - белые. Во втором ящике - 12 мячей, из которых 5 белых. Из первого ящика вытаскивают случайным образом два мяча, из второго - 4. Какова вероятность того, что 3 мяча белые?

In [ ]:
# Из первого ящика вытаскивают 1 белый шар, а из второго - 2 белых
p1 = (combinations(5, 1) * combinations(3, 1)) / combinations(8, 2) * (combinations(5, 2) * combinations(7, 2) / combinations(12, 4))
print(p1)
# Из первого ящика не вытаскивают белых мячей, а из второго - 3 белых и 1 чёрный
p2 = (combinations(5, 0) * combinations(3, 2)) / combinations(8, 2) * (combinations(5, 3) * combinations(7, 1)) / combinations(12, 4)
print(p2)
# Из первого ящика достают 2 белых, а из второго - 1 белый
p3 = (combinations(5, 2) * combinations(3, 0)) / combinations(8, 2) * (combinations(5, 1) * combinations(7, 3)) / combinations(12, 4)
print(p3)

0.22727272727272727
0.015151515151515152
0.12626262626262627


In [ ]:
# Полученные вероятности складываем, т.к. данные события не происходят одновременно
p = p1 + p2 + p3
print(round(p, 3))

0.369


**Ответ:** Вероятность того, что три мяча белые - 0,369

## Задание 3

На соревновании по биатлону один из трех спортсменов стреляет и попадает в мишень. Вероятность попадания для первого спортсмена равна 0.9, для второго — 0.8, для третьего — 0.6. Найти вероятность того, что выстрел произведен: a). первым спортсменом б). вторым спортсменом в). третьим спортсменом.

In [ ]:
# Вероятность того, что стреляет любой из спортсменов равна 1/3
# По формуле полной вероятности
P = 1/3 * (0.9 + 0.8 + 0.6)
print(P)

0.7666666666666667


In [ ]:
def bayes(p, P):
  return round((1/3 * p) / P, 3)

In [ ]:
# Применим формулу Байеса для каждого спортсмена
print(bayes(0.9, P))
print(bayes(0.8, P))
print(bayes(0.6, P))

0.391
0.348
0.261


**Ответ:** Вероятность того, что выстрел произведён первым спортсменом - 0.391, вторым - 0.348 и третьим - 0.261

## Задание 4

В университет на факультеты A и B поступило равное количество студентов, а на факультет C студентов поступило столько же, сколько на A и B вместе. Вероятность того, что студент факультета A сдаст первую сессию, равна 0.8. Для студента факультета B эта вероятность равна 0.7, а для студента факультета C - 0.9. Студент сдал первую сессию. Какова вероятность, что он учится: a). на факультете A б). на факультете B в). на факультете C?

In [ ]:
# вероятность того, что студент окажется с факультета А или B - 1/4, а с факультета С - 1/2
p1, p2 = 0.25, 0.5
# по формуле полной вероятности
P = p1 * 0.8 + p1 *0.7 + p2 * 0.9
print(P)

0.825


In [ ]:
def bayes(p, P, p0):
  return round((p0 * p) / P, 3)

In [ ]:
# рассчитаем вероятность сдачи экзамена для студента с каждого факультета
print(bayes(0.8, P, p1))
print(bayes(0.7, P, p1))
print(bayes(0.9, P, p2))


0.242
0.212
0.545


**Ответ:** Вероятность того, что студент учится на факультете А - 0.242, факультете В - 0.212, факультете С - 0.545

### Задание 5

Устройство состоит из трех деталей. Для первой детали вероятность выйти из строя в первый месяц равна 0.1, для второй - 0.2, для третьей - 0.25. Какова вероятность того, что в первый месяц выйдут из строя: а). все детали б). только две детали в). хотя бы одна деталь г). от одной до двух деталей?

In [ ]:
p1, p2, p3 = 0.1, 0.2, 0.25
# вероятность выхода всех деталей из строя
p_a = p1 * p2 * p3
print(round(p_a, 3))

0.005


In [ ]:
# вероятность выхода из строя двух любых деталей
p_b = p1 * p2 * (1 - p3) + p1 * p3 * (1 - p2) + p2 * p3 * (1 - p1)
print(round(p_b, 3))


0.08


In [ ]:
# вероятность выхода из строя хотя бы одной детали
p_c = 1 - (1 - p1) * (1 - p2) * (1 - p3)
print(round(p_c, 3))

0.46


In [ ]:
# вероятность выхода из строя от одной до двух деталей
p_d = 1 - (1 - p1) * (1 - p2)* (1 - p3) - p1 * p2 * p3
print(round(p_d, 3))

0.455


**Ответ:** вероятность выхода всех деталей из строя - 0.005, вероятность выхода из строя двух любых деталей - 0.08, вероятность выхода из строя хотя бы одной детали - 0.46, вероятность выхода из строя от одной до двух деталей 0,455